File: solutions/programmers/kit/knight_weapon.kt
fun solution(num : Int,
             limit : Int,
             power : Int) : Int {
    val divisors = IntArray(num +1) { getDivisorCount(it) }

    return divisors.slice(1..num).sumOf {
      if (it > limit) power else it
    }
}

fun getDivisorCount(num : Int) : Int {
    if (num ==0) return 0
    var count = 0
    var i = 1

    while (i * i <= num) {
        if (i*i == num) {
           count++
        } else if (num % i == 0) {
          count += 2
        }
        i++
    }

    return count

}


fun main() {
    println(solution(10, 3, 2))
    println(solution(5, 3, 2))
}

File: solutions/programmers/kit/knight_weapon.go
package main

import "fmt"

func solve(num int, limit int, power int) int {
	arr := []int{} // initalize an empty array
	res := 0

	// 약수의 개수 구하기

	for i := 1; i <= num; i++ {
		arr = append(arr, get_num_divisor(i))
	}

	fmt.Println(arr)

	process(arr, limit, power)

	for _, v := range arr {
		res += v
	}

	fmt.Println(res)

	return res
}

func process(arr []int, limit, power int) []int {
	for i, v := range arr {
		if v > limit {
			arr[i] = power
		}
	}
	return arr
}

func get_num_divisor(num int) int {
	count := 0
	for i := 1; i*i <= num; i++ {
		if i*i == num {
			count++
			continue
		}

		if num%i == 0 {
			count += 2
			continue
		}
	}

	return count
}

func main() {
	solve(5, 3, 2)
	solve(10, 3, 2)
}
